File: backend/app/engine/calculator.py
"""
Motor de cálculo RIC — NCh Elec 4/2003
Portado desde calculadora_ric_v7.html (JavaScript → Python)
Lógica idéntica al prototipo HTML: mismas fórmulas, misma tabla, mismos factores.
"""
import math
from typing import Optional

from app.engine.ric_tables import (
    TABLA_RIC, RicRow, SEC_MIN_CIRCUITO, K_CONST, FACTOR_FALLA,
    DIAMETRO_CABLE, get_temp_factor, get_grouping_factor,
    get_altitude_factor, is_air_installation, is_buried_installation,
)
from app.engine.schemas import (
    CalculatorInput, CalculatorResult, CalculatorResponse,
    RadioCurvatura, EstresTermico,
)
from app.engine.protection_selector import seleccionar_proteccion

# Límite de caída por tipo de circuito (RIC Art. 5.5.4)
LIMITE_CAIDA_CIRCUITO: dict[str, float] = {
    "alumbrado":      2.0,
    "fuerza":         3.0,
    "tomacorrientes": 3.0,
    "motor":          3.0,
    "alimentador":    3.0,
}


def _get_imax(row: RicRow, material: str, es_aire: bool) -> int:
    """Retorna la capacidad de corriente según material e instalación."""
    if material == "cu":
        return row.icu_a if es_aire else row.icu_d
    else:
        return row.ial_a if es_aire else row.ial_d


def _calc_caida(sistema: str, i_calc: float, r_ohm_per_m: float,
                longitud: float, tension: float) -> tuple[float, float]:
    """Calcula caída de tensión (V, %)."""
    if sistema == "trifasico":
        dv = math.sqrt(3) * i_calc * r_ohm_per_m * longitud
        dv_pct = (dv / tension) * 100
    elif sistema == "bifasico":
        dv = 2 * i_calc * r_ohm_per_m * longitud
        dv_pct = (dv / (2 * tension)) * 100
    else:  # monofasico
        dv = 2 * i_calc * r_ohm_per_m * longitud
        dv_pct = (dv / tension) * 100
    return dv, dv_pct


def _calc_radio_curvatura(sec_mm2: float, montaje: str, sistema: str, cables_por_fase: int) -> RadioCurvatura:
    """Radio mínimo de curvatura — NEC 300.34 / IEC 60364-5-52."""
    dims = DIAMETRO_CABLE.get(sec_mm2)
    if dims is None:
        d_uni = math.sqrt(sec_mm2) * 3.8 + 3
        d_multi = math.sqrt(sec_mm2) * 5.5 + 5
    else:
        d_uni, d_multi = dims

    es_unipolar = sistema != "monofasico"
    d_base = d_uni if es_unipolar else d_multi

    if montaje == "banco":
        factor = 12
        desc = "12× diámetro (banco de ducto)"
    elif montaje == "oculto":
        factor = 8
        desc = "8× diámetro (oculto/empotrado)"
    else:
        if es_unipolar:
            factor = 8
            desc = "8× diámetro (unipolar a la vista)"
        else:
            factor = 6
            desc = "6× diámetro (multiconductor a la vista)"

    radio = d_base * factor
    return RadioCurvatura(
        radio_mm=round(radio, 2),
        radio_interno_mm=round(radio - d_base, 2),
        diametro_mm=round(d_base, 2),
        factor=factor,
        descripcion_factor=desc,
        tipo_constructivo="Unipolar (THW)" if es_unipolar else "Multiconductor (THWN/NYY)",
        montaje=montaje,
    )


def _calc_estres_termico(
    sec_mm2: float, material: str, icc_ka: float, tiempo_s: float,
    tipo_falla: str, t_inicial: float, t_max: float
) -> EstresTermico:
    """Estrés térmico por cortocircuito — IEC 60949.
    Fórmula: S_min = (Icc_efectiva × √t) / K
    """
    factor_falla = FACTOR_FALLA.get(tipo_falla, 1.0)
    icc_efectiva_a = icc_ka * 1000 * factor_falla

    es_xlpe = t_max >= 200
    k_key = ("cu_" if material == "cu" else "al_") + ("xlpe" if es_xlpe else "thw")
    K = K_CONST[k_key]

    sec_min = (icc_efectiva_a * math.sqrt(tiempo_s)) / K
    icc_max_soportada = (sec_mm2 * K) / math.sqrt(tiempo_s) / 1000

    i2t = (icc_efectiva_a ** 2) * tiempo_s
    i2t_max = (sec_mm2 * K) ** 2
    ratio = i2t / i2t_max
    t_final = t_inicial + ratio * (t_max - t_inicial)

    return EstresTermico(
        icc_efectiva_ka=round(icc_efectiva_a / 1000, 3),
        factor_falla=factor_falla,
        tipo_falla=tipo_falla.upper(),
        k_const=K,
        es_xlpe=es_xlpe,
        sec_min_termica_mm2=math.ceil(sec_min * 10) / 10,
        icc_max_soportada_ka=round(icc_max_soportada, 3),
        i2t_ja=round(i2t, 0),
        i2t_max_ja=round(i2t_max, 0),
        ratio_saturacion=round(ratio, 4),
        t_final_estimada_c=round(t_final, 1),
        t_max_c=t_max,
        cumple=sec_mm2 >= sec_min,
    )


def _buscar_seccion(n_cables: int, i_req: float, material: str, es_aire: bool) -> Optional[RicRow]:
    """Busca la primera fila de TABLA_RIC que cumpla capacidad térmica."""
    for row in TABLA_RIC:
        imax = _get_imax(row, material, es_aire)
        if imax == 0:
            continue
        if imax * n_cables >= i_req:
            return row
    return None


def calculate(inp: CalculatorInput) -> CalculatorResponse:
    """
    Motor de cálculo principal RIC — portado 1:1 desde calculadora_ric_v7.html.

    Pasos:
    1. Corriente de diseño I_diseño
    2. Aplicar Fd → I_calc
    3. Factores Ft, Fg, Fa → factor_total
    4. I_req = I_calc / factor_total
    5. Buscar sección térmica (auto-paralelo hasta 3 cables)
    6. Sección mínima normativa RIC Art. 5.3.1
    7. Caída de tensión
    8. Ajustar sección si caída > límite
    9. Radio curvatura y estrés térmico
    """
    advertencias: list[str] = []

    # ── 1. Corriente de diseño ────────────────────────────────────────────────
    kw = inp.potencia_kw * 1000
    if inp.sistema == "trifasico":
        i_diseno = kw / (math.sqrt(3) * inp.tension_v * inp.factor_potencia)
    elif inp.sistema == "bifasico":
        i_diseno = kw / (2 * inp.tension_v * inp.factor_potencia)
    else:
        i_diseno = kw / (inp.tension_v * inp.factor_potencia)

    # ── 2. Factor de demanda ──────────────────────────────────────────────────
    i_calc = i_diseno * inp.factor_demanda

    # ── 3. Factores de corrección ─────────────────────────────────────────────
    ft = get_temp_factor(inp.temp_ambiente_c)
    fg = get_grouping_factor(inp.circuitos_agrupados)
    fa = get_altitude_factor(inp.msnm)
    factor_total = ft * fg * fa

    # ── 4. Corriente requerida (sin correcciones) ─────────────────────────────
    i_req = i_calc / factor_total

    es_aire = is_air_installation(inp.tipo_canalizacion)
    es_enterrado = is_buried_installation(inp.tipo_canalizacion)

    # ── 5. Buscar sección por capacidad térmica ───────────────────────────────
    fila_elegida: Optional[RicRow] = None
    cables_por_fase = 1

    if inp.cables_por_fase == 0:
        for n in range(1, 4):
            fila_elegida = _buscar_seccion(n, i_req, inp.material, es_aire)
            if fila_elegida:
                cables_por_fase = n
                break
    else:
        cables_por_fase = inp.cables_por_fase
        fila_elegida = _buscar_seccion(cables_por_fase, i_req, inp.material, es_aire)

    if fila_elegida is None:
        raise ValueError(
            "La corriente calculada supera el máximo de las tablas RIC disponibles. "
            "Se requiere media tensión o análisis especial."
        )

    # ── 5b. Override manual de sección (desde editor visual) ─────────────────
    if inp.seccion_forzada_mm2 is not None:
        fila_forzada = next((r for r in TABLA_RIC if r.sec == inp.seccion_forzada_mm2), None)
        if fila_forzada:
            fila_elegida = fila_forzada
            advertencias.append(f"Sección ingresada manualmente: {inp.seccion_forzada_mm2} mm²")

    # ── 6. Sección mínima normativa (RIC Art. 5.3.1) ─────────────────────────
    sec_min_ric = SEC_MIN_CIRCUITO.get(inp.tipo_circuito, 1.5)
    ajustado_por_minimo = False
    if fila_elegida.sec < sec_min_ric:
        fila_min = next((r for r in TABLA_RIC if r.sec >= sec_min_ric), None)
        if fila_min:
            fila_elegida = fila_min
            ajustado_por_minimo = True
    elif fila_elegida.sec == sec_min_ric:
        # Si la corriente de diseño es menor al 70% de la capacidad de la sección
        # mínima, el mínimo normativo es el criterio vinculante (no la térmica)
        imax_min = _get_imax(fila_elegida, inp.material, es_aire) * factor_total
        if imax_min > 0 and i_req < imax_min * 0.7:
            ajustado_por_minimo = True

    seccion_elegida = fila_elegida.sec
    r_cu = inp.material == "cu"
    resistencia_ohm_km = fila_elegida.rcu if r_cu else fila_elegida.ral
    r_ohm_per_m = resistencia_ohm_km / (cables_por_fase * 1000)

    # ── 7. Caída de tensión ───────────────────────────────────────────────────
    dv, dv_pct = _calc_caida(inp.sistema, i_calc, r_ohm_per_m, inp.longitud_m, inp.tension_v)
    lim_caida = inp.limite_caida_pct if inp.limite_caida_pct else LIMITE_CAIDA_CIRCUITO.get(inp.tipo_circuito, 3.0)

    # ── 8. Ajustar sección si caída excede límite ─────────────────────────────
    sec_caida = seccion_elegida
    fila_caida = fila_elegida
    cables_caida = cables_por_fase
    ajustado_por_caida = False

    if dv_pct > lim_caida:
        outer_break = False
        for row in TABLA_RIC:
            if row.sec < seccion_elegida:  # solo secciones >= térmica
                continue
            for n in range(cables_por_fase, 4):
                r = (row.rcu if r_cu else row.ral) / (n * 1000)
                _, dvpct_test = _calc_caida(inp.sistema, i_calc, r, inp.longitud_m, inp.tension_v)
                imax_test = _get_imax(row, inp.material, es_aire)
                if imax_test == 0:
                    continue
                if dvpct_test <= lim_caida and imax_test * n * factor_total >= i_calc:
                    if row.sec > seccion_elegida or n > cables_por_fase:
                        sec_caida = row.sec
                        fila_caida = row
                        cables_caida = n
                        ajustado_por_caida = True
                    outer_break = True
                    break
            if outer_break:
                break

    # Seleccionar la mayor entre sección térmica y sección por caída
    if sec_caida >= seccion_elegida:
        fila_final = fila_caida
        cables_final = cables_caida
    else:
        fila_final = fila_elegida
        cables_final = cables_por_fase

    sec_final = fila_final.sec

    # ── Caída final ───────────────────────────────────────────────────────────
    r_final = (fila_final.rcu if r_cu else fila_final.ral) / (cables_final * 1000)
    dv_final, dv_pct_final = _calc_caida(inp.sistema, i_calc, r_final, inp.longitud_m, inp.tension_v)

    imax_final = _get_imax(fila_final, inp.material, es_aire) * cables_final * factor_total

    cumple_termico = imax_final >= i_calc
    cumple_caida = dv_pct_final <= lim_caida
    cumple = cumple_termico and cumple_caida

    # ── 9. Conductores neutro y tierra ────────────────────────────────────────
    if inp.sistema == "trifasico" and sec_final > 16:
        sec_neutro = sec_final / 2
    else:
        sec_neutro = sec_final

    if sec_final <= 16:
        sec_tierra = sec_final
    elif sec_final <= 35:
        sec_tierra = 16.0
    else:
        sec_tierra = sec_final / 2

    # Descripción de configuración
    if inp.sistema == "trifasico":
        neutro_desc = f"{math.ceil(cables_final/2)} cable(s) neutro" if cables_final > 1 else "1 cable neutro"
        descripcion_config = f"{cables_final} cable(s)/fase × 3 fases + {neutro_desc} + tierra"
    elif inp.sistema == "bifasico":
        descripcion_config = f"{cables_final} cable(s)/fase × 2 fases + {cables_final} cable(s) neutro + tierra"
    else:
        descripcion_config = f"{cables_final} cable(s) fase + {cables_final} cable(s) neutro + tierra"

    # ── 10. Radio de curvatura ────────────────────────────────────────────────
    radio_data = _calc_radio_curvatura(sec_final, inp.montaje, inp.sistema, cables_final)

    # ── 11. Estrés térmico ────────────────────────────────────────────────────
    estres_data: Optional[EstresTermico] = None
    if inp.icc_ka and inp.tiempo_cc_s:
        estres_data = _calc_estres_termico(
            sec_final, inp.material, inp.icc_ka, inp.tiempo_cc_s,
            inp.tipo_falla, inp.t_inicial_c, inp.t_max_c
        )

    # ── Advertencias ──────────────────────────────────────────────────────────
    if ajustado_por_minimo:
        advertencias.append(
            f"Sección aumentada al mínimo normativo RIC Art. 5.3.1: {sec_min_ric} mm² "
            f"para circuito tipo '{inp.tipo_circuito}'."
        )
    if ajustado_por_caida:
        advertencias.append(
            f"Sección aumentada por criterio caída de tensión. "
            f"ΔV = {round(dv_pct_final, 2)}% vs. límite {lim_caida}%."
        )
    if inp.msnm > 1000:
        reduccion = round((1 - fa) * 100, 0)
        advertencias.append(
            f"Altitud {inp.msnm} msnm: Fa = {fa} — capacidad reducida {reduccion}% (IEC 60364-5-52)."
        )
    if es_enterrado:
        advertencias.append(
            "Instalación enterrada: verificar profundidad mínima RIC Art. 5.4 y arena de relleno."
        )
    if cables_final > 1:
        advertencias.append(
            "Paralelo: cables de igual longitud, material, sección y tipo (RIC Art. 5.3.2)."
        )

    # Selección de protecciones (termomagnético + diferencial)
    proteccion_data = seleccionar_proteccion(
        i_b=round(i_calc, 3),
        i_z=round(imax_final, 3),
        tipo_circuito=inp.tipo_circuito,
        sistema=inp.sistema,
        icc_ka=inp.icc_ka,
    )

    resultado = CalculatorResult(
        seccion_mm2=sec_final,
        calibre_awg=fila_final.awg,
        cables_por_fase=cables_final,
        material=inp.material,
        i_diseno_a=round(i_diseno, 3),
        i_calc_a=round(i_calc, 3),
        i_req_a=round(i_req, 3),
        i_max_corregida_a=round(imax_final, 3),
        caida_pct=round(dv_pct_final, 4),
        caida_v=round(dv_final, 4),
        limite_caida_pct=lim_caida,
        cumple_termico=cumple_termico,
        cumple_caida=cumple_caida,
        cumple=cumple,
        ft=ft,
        fg=fg,
        fa=fa,
        factor_total=round(factor_total, 4),
        sec_neutro_mm2=sec_neutro,
        sec_tierra_mm2=sec_tierra,
        descripcion_config=descripcion_config,
        radio_curvatura=radio_data,
        estres_termico=estres_data,
        proteccion=proteccion_data,
        ajustado_por_minimo=ajustado_por_minimo,
        ajustado_por_caida=ajustado_por_caida,
        sec_min_ric_mm2=sec_min_ric,
        advertencias=advertencias,
    )

    return CalculatorResponse(ok=True, resultado=resultado, advertencias=advertencias)

File: frontend/src/lib/engine/calculator.ts
/**
 * Motor de cálculo RIC — portado 1:1 desde backend/app/engine/calculator.py
 *
 * Mismas fórmulas, misma tabla, mismos factores. Diseñado para correr offline
 * en el browser / Capacitor sin necesidad de backend.
 */
import {
  TABLA_RIC, SEC_MIN_CIRCUITO, K_CONST, FACTOR_FALLA, DIAMETRO_CABLE,
  getTempFactor, getGroupingFactor, getAltitudeFactor,
  isAirInstallation, isBuriedInstallation,
  type RicRow,
} from './ric-tables'
import type {
  CalculatorInput, CalculatorResult, CalculatorResponse,
  RadioCurvatura, EstresTermico,
} from './types'
import { seleccionarProteccion } from './protection-selector'

// ── Límite de caída por tipo de circuito (RIC Art. 5.5.4) ────────────────────

const LIMITE_CAIDA_CIRCUITO: Record<string, number> = {
  alumbrado:      2.0,
  fuerza:         3.0,
  tomacorrientes: 3.0,
  motor:          3.0,
  alimentador:    3.0,
}

// ── Helpers internos ─────────────────────────────────────────────────────────

function round(v: number, dec: number): number {
  const f = 10 ** dec
  return Math.round(v * f) / f
}

function getImax(row: RicRow, material: string, esAire: boolean): number {
  if (material === 'cu') return esAire ? row.icu_a : row.icu_d
  return esAire ? row.ial_a : row.ial_d
}

function calcCaida(
  sistema: string, iCalc: number, rOhmPerM: number, longitud: number, tension: number,
): [number, number] {
  let dv: number
  let dvPct: number
  if (sistema === 'trifasico') {
    dv = Math.sqrt(3) * iCalc * rOhmPerM * longitud
    dvPct = (dv / tension) * 100
  } else if (sistema === 'bifasico') {
    dv = 2 * iCalc * rOhmPerM * longitud
    dvPct = (dv / (2 * tension)) * 100
  } else {
    dv = 2 * iCalc * rOhmPerM * longitud
    dvPct = (dv / tension) * 100
  }
  return [dv, dvPct]
}

function calcRadioCurvatura(
  secMm2: number, montaje: string, sistema: string, _cablesPorFase: number,
): RadioCurvatura {
  const dims = DIAMETRO_CABLE[secMm2]
  let dUni: number, dMulti: number
  if (dims == null) {
    dUni = Math.sqrt(secMm2) * 3.8 + 3
    dMulti = Math.sqrt(secMm2) * 5.5 + 5
  } else {
    ;[dUni, dMulti] = dims
  }

  const esUnipolar = sistema !== 'monofasico'
  const dBase = esUnipolar ? dUni : dMulti

  let factor: number
  let desc: string
  if (montaje === 'banco') {
    factor = 12; desc = '12× diámetro (banco de ducto)'
  } else if (montaje === 'oculto') {
    factor = 8; desc = '8× diámetro (oculto/empotrado)'
  } else if (esUnipolar) {
    factor = 8; desc = '8× diámetro (unipolar a la vista)'
  } else {
    factor = 6; desc = '6× diámetro (multiconductor a la vista)'
  }

  const radio = dBase * factor
  return {
    radio_mm: round(radio, 2),
    radio_interno_mm: round(radio - dBase, 2),
    diametro_mm: round(dBase, 2),
    factor,
    descripcion_factor: desc,
    tipo_constructivo: esUnipolar ? 'Unipolar (THW)' : 'Multiconductor (THWN/NYY)',
    montaje,
  }
}

function calcEstresTermico(
  secMm2: number, material: string, iccKa: number, tiempoS: number,
  tipoFalla: string, tInicial: number, tMax: number,
): EstresTermico {
  const factorFalla = FACTOR_FALLA[tipoFalla] ?? 1.0
  const iccEfectivaA = iccKa * 1000 * factorFalla

  const esXlpe = tMax >= 200
  const kKey = (material === 'cu' ? 'cu_' : 'al_') + (esXlpe ? 'xlpe' : 'thw')
  const K = K_CONST[kKey]

  const secMin = (iccEfectivaA * Math.sqrt(tiempoS)) / K
  const iccMaxSoportada = (secMm2 * K) / Math.sqrt(tiempoS) / 1000

  const i2t = iccEfectivaA ** 2 * tiempoS
  const i2tMax = (secMm2 * K) ** 2
  const ratio = i2t / i2tMax
  const tFinal = tInicial + ratio * (tMax - tInicial)

  return {
    icc_efectiva_ka: round(iccEfectivaA / 1000, 3),
    factor_falla: factorFalla,
    tipo_falla: tipoFalla.toUpperCase(),
    k_const: K,
    es_xlpe: esXlpe,
    sec_min_termica_mm2: Math.ceil(secMin * 10) / 10,
    icc_max_soportada_ka: round(iccMaxSoportada, 3),
    i2t_ja: Math.round(i2t),
    i2t_max_ja: Math.round(i2tMax),
    ratio_saturacion: round(ratio, 4),
    t_final_estimada_c: round(tFinal, 1),
    t_max_c: tMax,
    cumple: secMm2 >= secMin,
  }
}

function buscarSeccion(
  nCables: number, iReq: number, material: string, esAire: boolean,
): RicRow | null {
  for (const row of TABLA_RIC) {
    const imax = getImax(row, material, esAire)
    if (imax === 0) continue
    if (imax * nCables >= iReq) return row
  }
  return null
}

// ── Función principal ─────────────────────────────────────────────────────────

export function calculate(inp: CalculatorInput): CalculatorResponse {
  const advertencias: string[] = []

  // ── 1. Corriente de diseño ──────────────────────────────────────────────
  const kw = inp.potencia_kw * 1000
  let iDiseno: number
  if (inp.sistema === 'trifasico') {
    iDiseno = kw / (Math.sqrt(3) * inp.tension_v * inp.factor_potencia)
  } else if (inp.sistema === 'bifasico') {
    iDiseno = kw / (2 * inp.tension_v * inp.factor_potencia)
  } else {
    iDiseno = kw / (inp.tension_v * inp.factor_potencia)
  }

  // ── 2. Factor de demanda ────────────────────────────────────────────────
  const iCalc = iDiseno * inp.factor_demanda

  // ── 3. Factores de corrección ───────────────────────────────────────────
  const ft = getTempFactor(inp.temp_ambiente_c)
  const fg = getGroupingFactor(inp.circuitos_agrupados)
  const fa = getAltitudeFactor(inp.msnm)
  const factorTotal = ft * fg * fa

  // ── 4. Corriente requerida ──────────────────────────────────────────────
  const iReq = iCalc / factorTotal

  const esAire = isAirInstallation(inp.tipo_canalizacion)
  const esEnterrado = isBuriedInstallation(inp.tipo_canalizacion)

  // ── 5. Buscar sección por capacidad térmica ─────────────────────────────
  let filaElegida: RicRow | null = null
  let cablesPorFase = 1

  if (inp.cables_por_fase === 0) {
    for (let n = 1; n <= 3; n++) {
      filaElegida = buscarSeccion(n, iReq, inp.material, esAire)
      if (filaElegida) { cablesPorFase = n; break }
    }
  } else {
    cablesPorFase = inp.cables_por_fase
    filaElegida = buscarSeccion(cablesPorFase, iReq, inp.material, esAire)
  }

  if (filaElegida == null) {
    throw new Error(
      'La corriente calculada supera el máximo de las tablas RIC disponibles. ' +
      'Se requiere media tensión o análisis especial.'
    )
  }

  // ── 5b. Override manual ─────────────────────────────────────────────────
  if (inp.seccion_forzada_mm2 != null) {
    const filaForzada = TABLA_RIC.find(r => r.sec === inp.seccion_forzada_mm2)
    if (filaForzada) {
      filaElegida = filaForzada
      advertencias.push(`Sección ingresada manualmente: ${inp.seccion_forzada_mm2} mm²`)
    }
  }

  // ── 6. Sección mínima normativa (RIC Art. 5.3.1) ───────────────────────
  const secMinRic = SEC_MIN_CIRCUITO[inp.tipo_circuito] ?? 1.5
  let ajustadoPorMinimo = false

  if (filaElegida.sec < secMinRic) {
    const filaMin = TABLA_RIC.find(r => r.sec >= secMinRic)
    if (filaMin) { filaElegida = filaMin; ajustadoPorMinimo = true }
  } else if (filaElegida.sec === secMinRic) {
    const imaxMin = getImax(filaElegida, inp.material, esAire) * factorTotal
    if (imaxMin > 0 && iReq < imaxMin * 0.7) ajustadoPorMinimo = true
  }

  let seccionElegida = filaElegida.sec
  const rCu = inp.material === 'cu'
  const resistenciaOhmKm = rCu ? filaElegida.rcu : filaElegida.ral
  const rOhmPerM = resistenciaOhmKm / (cablesPorFase * 1000)

  // ── 7. Caída de tensión ─────────────────────────────────────────────────
  const [_dv, dvPct] = calcCaida(inp.sistema, iCalc, rOhmPerM, inp.longitud_m, inp.tension_v)
  const limCaida = inp.limite_caida_pct ?? LIMITE_CAIDA_CIRCUITO[inp.tipo_circuito] ?? 3.0

  // ── 8. Ajustar sección si caída excede límite ──────────────────────────
  let secCaida = seccionElegida
  let filaCaida = filaElegida
  let cablesCaida = cablesPorFase
  let ajustadoPorCaida = false

  if (dvPct > limCaida) {
    let outerBreak = false
    for (const row of TABLA_RIC) {
      if (row.sec < seccionElegida) continue
      for (let n = cablesPorFase; n <= 3; n++) {
        const r = (rCu ? row.rcu : row.ral) / (n * 1000)
        const [, dvPctTest] = calcCaida(inp.sistema, iCalc, r, inp.longitud_m, inp.tension_v)
        const imaxTest = getImax(row, inp.material, esAire)
        if (imaxTest === 0) continue
        if (dvPctTest <= limCaida && imaxTest * n * factorTotal >= iCalc) {
          if (row.sec > seccionElegida || n > cablesPorFase) {
            secCaida = row.sec; filaCaida = row; cablesCaida = n; ajustadoPorCaida = true
          }
          outerBreak = true; break
        }
      }
      if (outerBreak) break
    }
  }

  // Seleccionar la mayor entre sección térmica y sección por caída
  let filaFinal: RicRow
  let cablesFinal: number
  if (secCaida >= seccionElegida) {
    filaFinal = filaCaida; cablesFinal = cablesCaida
  } else {
    filaFinal = filaElegida; cablesFinal = cablesPorFase
  }

  const secFinal = filaFinal.sec

  // ── Caída final ─────────────────────────────────────────────────────────
  const rFinal = (rCu ? filaFinal.rcu : filaFinal.ral) / (cablesFinal * 1000)
  const [dvFinal, dvPctFinal] = calcCaida(inp.sistema, iCalc, rFinal, inp.longitud_m, inp.tension_v)
  const imaxFinal = getImax(filaFinal, inp.material, esAire) * cablesFinal * factorTotal

  const cumpleTermico = imaxFinal >= iCalc
  const cumpleCaida = dvPctFinal <= limCaida
  const cumple = cumpleTermico && cumpleCaida

  // ── 9. Conductores neutro y tierra ──────────────────────────────────────
  let secNeutro: number
  if (inp.sistema === 'trifasico' && secFinal > 16) {
    secNeutro = secFinal / 2
  } else {
    secNeutro = secFinal
  }

  let secTierra: number
  if (secFinal <= 16) secTierra = secFinal
  else if (secFinal <= 35) secTierra = 16.0
  else secTierra = secFinal / 2

  // Descripción de configuración
  let descripcionConfig: string
  if (inp.sistema === 'trifasico') {
    const neutroDesc = cablesFinal > 1
      ? `${Math.ceil(cablesFinal / 2)} cable(s) neutro`
      : '1 cable neutro'
    descripcionConfig = `${cablesFinal} cable(s)/fase × 3 fases + ${neutroDesc} + tierra`
  } else if (inp.sistema === 'bifasico') {
    descripcionConfig = `${cablesFinal} cable(s)/fase × 2 fases + ${cablesFinal} cable(s) neutro + tierra`
  } else {
    descripcionConfig = `${cablesFinal} cable(s) fase + ${cablesFinal} cable(s) neutro + tierra`
  }

  // ── 10. Radio de curvatura ──────────────────────────────────────────────
  const radioData = calcRadioCurvatura(secFinal, inp.montaje, inp.sistema, cablesFinal)

  // ── 11. Estrés térmico ──────────────────────────────────────────────────
  let estresData: EstresTermico | null = null
  if (inp.icc_ka && inp.tiempo_cc_s) {
    estresData = calcEstresTermico(
      secFinal, inp.material, inp.icc_ka, inp.tiempo_cc_s,
      inp.tipo_falla, inp.t_inicial_c, inp.t_max_c,
    )
  }

  // ── Advertencias ────────────────────────────────────────────────────────
  if (ajustadoPorMinimo) {
    advertencias.push(
      `Sección aumentada al mínimo normativo RIC Art. 5.3.1: ${secMinRic} mm² ` +
      `para circuito tipo '${inp.tipo_circuito}'.`
    )
  }
  if (ajustadoPorCaida) {
    advertencias.push(
      `Sección aumentada por criterio caída de tensión. ` +
      `ΔV = ${round(dvPctFinal, 2)}% vs. límite ${limCaida}%.`
    )
  }
  if (inp.msnm > 1000) {
    const reduccion = Math.round((1 - fa) * 100)
    advertencias.push(
      `Altitud ${inp.msnm} msnm: Fa = ${fa} — capacidad reducida ${reduccion}% (IEC 60364-5-52).`
    )
  }
  if (esEnterrado) {
    advertencias.push('Instalación enterrada: verificar profundidad mínima RIC Art. 5.4 y arena de relleno.')
  }
  if (cablesFinal > 1) {
    advertencias.push('Paralelo: cables de igual longitud, material, sección y tipo (RIC Art. 5.3.2).')
  }

  // Selección de protecciones
  const proteccionData = seleccionarProteccion(
    round(iCalc, 3),
    round(imaxFinal, 3),
    inp.tipo_circuito,
    inp.sistema,
    inp.icc_ka,
  )

  const resultado: CalculatorResult = {
    seccion_mm2: secFinal,
    calibre_awg: filaFinal.awg,
    cables_por_fase: cablesFinal,
    material: inp.material,
    i_diseno_a: round(iDiseno, 3),
    i_calc_a: round(iCalc, 3),
    i_req_a: round(iReq, 3),
    i_max_corregida_a: round(imaxFinal, 3),
    caida_pct: round(dvPctFinal, 4),
    caida_v: round(dvFinal, 4),
    limite_caida_pct: limCaida,
    cumple_termico: cumpleTermico,
    cumple_caida: cumpleCaida,
    cumple,
    ft,
    fg,
    fa,
    factor_total: round(factorTotal, 4),
    sec_neutro_mm2: secNeutro,
    sec_tierra_mm2: secTierra,
    descripcion_config: descripcionConfig,
    radio_curvatura: radioData,
    estres_termico: estresData,
    proteccion: proteccionData,
    ajustado_por_minimo: ajustadoPorMinimo,
    ajustado_por_caida: ajustadoPorCaida,
    sec_min_ric_mm2: secMinRic,
    advertencias,
  }

  return { ok: true, resultado, advertencias }
}
